addons/portal/models/mail_thread.py

```diff
@@ -1,10 +1,32 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 import hashlib
 import hmac
 
-from odoo import fields, models, _
+from werkzeug.exceptions import NotFound
+
+from odoo import api, fields, models, _
+from odoo.tools import frozendict
+from odoo.addons.portal.controllers.mail import _check_special_access
+
+
+def _check_thread_portal_access(env, thread_model, thread_id, token, _hash, pid):
+    has_access = False
+    if token or (_hash and pid):
+        pid = int(pid) if pid else False
+        has_access = _check_special_access(thread_model, thread_id, token, _hash, pid)
+        if has_access:
+            partner_id = env.user.partner_id
+            record = env[thread_model].sudo().browse(thread_id)
+            if _hash and pid:
+                partner_id = env["res.partner"].sudo().browse(pid)
+            elif token:
+                if env.user._is_public() and hasattr(record, 'partner_id') and record.partner_id:
+                    partner_id = record.partner_id
+                elif not partner_id:
+                    raise NotFound()
+            env.context = frozendict({**env.context, "portal_partner": partner_id})
+    return has_access
```

Comment:
We discussed about keeping the existing `_check_special_access` method in this PR rather than making a copy, can't this be done?

I don't want a copy, either the old method should be used, or the rest of the code should use the new method.

I prefer if we do the latter, as the old method is "locked" to a specific controller, but this one can be used anywhere.

Otherwise it gets really hard to maintain and the tests no longer guarantee that the method is correct (only one of them, not being sure which one).

Reply:
Note: if the method is moved here, it would be best if it was using thread as record set, to avoid having to pass thread_model and thread_id around.

Reply:
I removed the `_check_special_access` and `_message_post_helper`. `_check_thread_portal_access` is moved inside the thread and all the security check is happening there. So `res_model` is not in the parameter anymore.